hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/NodeDecommissionMetrics.java

```diff
@@ -135,10 +135,13 @@ public long getUnhealthyContainers() {
 
   private Map<String, ContainerStateInWorkflow> metricsByHost;
 
+  private Map<String, Long> startTimeByHost;
+
   /** Private constructor. */
   private NodeDecommissionMetrics() {
     this.registry = new MetricsRegistry(METRICS_SOURCE_NAME);
     metricsByHost = new HashMap<>();
+    startTimeByHost = new HashMap<>();
```

Comment:
All the host specific metrics are stored in `private Map<String, ContainerStateInWorkflow> metricsByHost;` right now. Could we extend ContainerStateInWorkflow to also carry the start time and use the same structure for storing this?

Then you could also add a static constant for the MetricsInfo, like the existing ones:

```
private static final MetricsInfo HOST_SUFFICIENTLY_REPLICATED = Interns
        .info(
            "SufficientlyReplicatedDN",
        "Number of sufficiently replicated containers "
            + "for host in decommissioning and "
            + "maintenance mode");
```

Then you would just need to add a getter for `getWorkflowStartTimeByHost` to retrieve the value.

Reply:
The `metricsByHost` which stores host specific metrics is cleared and repopulated during each run of the `DatanodeAdminMonitor`. We can see in the below method responsible for setting `metricsByHost` in `NodeDecommissionMetrics`:
```
public synchronized void metricRecordOfContainerStateByHost(
      Map<String, ContainerStateInWorkflow> containerStatesByHost) {
    metricsByHost.clear();
    metricsByHost.putAll(containerStatesByHost);
  }
```
But the start-time of decommissioning for a DN doesn't need to be set to a new value in each iteration of the `DatanodeAdminMonitor`. That is why I created a different map to store this information. 
I think if we want to store it in `ContainerStateInWorkflow`, we will have to temporarily store the start-time of each host somewhere in metricRecordOfContainerStateByHost method before clearing metricsByHost and then reassign it to the same values.
Is there any other way?

Reply:
One way, would be to modify:

```
private Set<DatanodeDetails> trackedNodes = new HashSet<>();
```
To be a set of TrackedDatanode, where TrackedDatanode is a public inner class within DatanodeAdminMonitorImpl, where it wraps the datanodeDetails and the start time, along with potentially other fields - eg, we could also store the counters we are passing to the metric in there if we wanted, but to start with, it would make sense to just try the startTime and see how it looks, then we could expand it if needed.

Then when forming the object to pass to the metrics on each iteration we can pull the details from there.

As I type this, I looked at the ContainerStateInWorkflow class, which is nearly the same thing in the metrics class, only it gets cleared each time.

If we created a TrackedDatanode object, we may be able to simplify away the use of the `private Map<String, ContainerStateInWorkflow>` in DatanodeAdminMonitor, but I would start small and see if the TrackedNode object to track the start time makes sense to begin with.

Reply:
An advantage of holding all the relevant details in trackedNodes, would be that to display these on the client, you don't even need to go to the metrics class and call individual getters - you could just return the trackedNodes set to the client, and it can display whatever is in it. Then the details are sent to the metrics as now to allow them to be displayed via metrics utilities.

Reply:
I have tried out this approach in this PR: https://github.com/apache/ozone/pull/5798. It stores start-time and datanodeDetails in one object, please do review it. Thank you!